### DISCUSSION TITLE
Most data types in API docs are "array[string]" in latest versions?

### DISCUSSION THREAD

**Discussion by bluikko** [netbox-community/netbox#13179] [Q&A] [answered]
In earlier versions of NetBox, probably before the major API change, data types were "string" which seemed normal.

In the latest version such as the demo site, at https://demo.netbox.dev/api/schema/swagger-ui/#/ipam/ipam_ip_addresses_list most of the data types are now "array[string]" which does not seem right to me.

Does anyone know if this is intentional or a bug? Can open an issue if it's really a bug.

**Comment by candlerb** [comment, marked as answer]
Those are *query parameters*, and I'd say it's correct that they are `array[]` because they can have repeated values.

e.g. `/api/ipam/ip-addresses/?device_id=1&device_id=2` shows all IP addresses which are attached to device_id 1 *or* device_id 2. Querying `?tag=foo&tag=bar` shows all those with tag `foo` *and* tag `bar`. Info [here](https://docs.netbox.dev/en/stable/reference/filtering/#filtering-objects).

SInce query parameters can be repeated, most libraries represent them as arrays, for example python's own [parse_qs](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.parse_qs):

> Data are returned as a dictionary. The dictionary keys are the unique query variable names and the values are *lists of values for each name.* [my emphasis]

Or in the opposite direction:

```
>>> import urllib.parse
>>> urllib.parse.urlencode({"device_id":[1,2],"tag":["foo","bar"]}, doseq=True)
'device_id=1&device_id=2&tag=foo&tag=bar'
```

**Reply by bluikko** [reply]
So it sounds like the new API docs is more accurate or just work differently from the old Swagger UI and the old behavior was inaccurate. Thanks, will mark that as the answer then.

**Reply by candlerb** [reply]
Yeah, there was a big change from OpenAPI 2 to OpenAPI 3 in Netbox 3.5.
